jmolecules-events/src/main/java/org/jmolecules/event/annotation/DomainEvent.java

```diff
@@ -29,4 +29,11 @@
  *      Reference (Evans) - Domain Events</a>
  */
 public @interface DomainEvent {
+
+    /**
+     * An identifier for the type of this event. This information may be used by external tools 
+     * to create a linkage between the event and the consumer.
+     */    
+    String type() default "";
```

Comment:
Wouldn't that connection be given implicitly by the type annotated with `@DomainEvent` and the parameter type of any method annotated with `@DomainEventHandler`?

Reply:
That only applies when the publisher and subscriber both reference the same DomainEvent class of a provided library. In case that there is no shared library for this (publisher and subscriber have their own impl), it wouldn't match. But maybe this thought is to theoretical.

Reply:
The fully qualified class name could be a good default if the "type" attribute is left unspecified.

We have a few lessons learned in Axon here. We are going to add ways to explicitly "type" events, rather than relying on their fqcn. While the name of the Event is generally not under discussion, the "namespace" is. There is often a request to decouple the namespace from the package name of a class.
The background is as follows:
a component that emits an event, publishes information within that event that it may deem relevant to describe it. However, a handler of an event may not be interested in all that information. In the "context" of the handler, only certain elements are relevant. You will want to express that in a class using a package that identifies that "context" of the handler. But the namespace of the event you handle, is still the one defined by the publisher.

Very likely, we will be adding stereotypes in Axon Framework, allowing developers to override the namespace (or maybe we're going to use the term "context").
There will be likely 2 attributes:
- context/namespace -> to override the use of the package name as the namespace, leaving the Simple Class Name as the name of the event.
- type/name -> to override the fully qualified type of the event, in case not only the namespace/context, but also the class name should be overridden.

Reply:
Alright, that's great input. If the fundamental idea is to abstract from the type system, I'd vote to move away from using `type` and specify the attributes as follows:

* `namespace` – Tooling should default this to the package name of the annotated type. Used to group events and allow clients to express interest in a certain set of events.
* `name` – Tooling should default this to the simple class name of the annotated type. Used to fix the actual name to guard it against refactorings.

The difference to what Allard suggested is that it think it's better to have symmetry in the attributes? If we wanted to follow Allard's suggestion, what would be the semantics of both the `namespace` and `name` attribute set to potentially conflicting values.

Reply:
Good points and also an interesting idea to split into namespace and name. I'll adapt this and add a note to the javadoc on how this might be used by tools.